Absensi read xlsx

riencus opened the pull request:
Pembuatan method untuk membaca file input berupa Excel, belum termasuk unit test, manual tested.
Sekaligus pemecahan method pembacaan file dari `EmployeeTappingService` ke `FileReaderComponent`

riencus committed:
Separate reading File from attendance service

Pembuatan interface FileReaderComponent dan Implementasinya untuk
mengubah input yang berupa file menjadi List<String>
Jika ditambahkan method readXlsx apakah attendance service nanti tidak
menjadi terlalu gemuk?

riencus committed:
Create ReadXlsx Method, no unit test

riencus committed:
Create readXls Method for excel 2007 and lower

riencus committed:
Create readXls Method for excel 2007 and lower

riencus committed:
Fix typo

- no return in FileReaderComponentImpl
- Override instead of Autowire

adhikasp commented on src/main/java/com/blibli/future/service/EmployeeTappingServiceImpl.java:
```diff
@@ -3,20 +3,19 @@
+import java.util.List;
+import java.util.Map;
 
 @Service //Question: Is it normal for all method in a service to return a boolean?
```
No it is not

adhikasp commented on src/main/java/com/blibli/future/service/FileReaderComponentImpl.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
```
Why use component, not service?

http://stackoverflow.com/a/23821409

riencus replied:
Hmm... That's right. Why did I write it as Component? Thanks for pointing it out, I'll change it.

adhikasp commented on src/main/java/com/blibli/future/service/FileReaderComponentImpl.java:
```diff
@@ -0,0 +1,87 @@
+@Component
+public class FileReaderComponentImpl implements FileReaderComponent {
+    @Override
+    public List<String> readFileAsStrings(MultipartFile file) {//Question: Is it Okay to do this? A class just for parsing a file?
```
It's already a pretty complicated class, so I think it's OK

adhikasp commented on src/main/java/com/blibli/future/service/FileReaderComponentImpl.java:
```diff
@@ -0,0 +1,87 @@
+        } else if(filename.endsWith(".xls")) {
+            return readXlsAsList(file);
+        } else {
+            return null; //or throw "Unsupported file"
```
IMO better throw an exception than silently fail.
https://softwareengineering.stackexchange.com/a/190531

adhikasp commented on src/main/java/com/blibli/future/service/FileReaderComponentImpl.java:
```diff
@@ -0,0 +1,87 @@
+    public List<String> readFileAsStrings(MultipartFile file) {//Question: Is it Okay to do this? A class just for parsing a file?
+        String filename = file.getContentType();
+
+        if(filename.endsWith(".csv")) {
```
Convert it first to lowercase/uppercase to avoid case sensitive matching. `something.CSV` doen't match here

adhikasp commented on src/main/java/com/blibli/future/service/EmployeeTappingServiceImpl.java:
```diff
@@ -167,22 +171,20 @@ public boolean addTapData(List<TapData> tapDataList) {
-            inputData = readCsvAsList(tapMachineFile);
-        } else {
-            return false; //Or should it throws some "Unsupported file Type" ?
+        if(inputData == null) {
```
If you throw exception while reading the file (in below) you became forced to handle it. No surprise result when you *forget* to handle special case / broken input.

riencus replied:
I'm sorry, I failed to understand your suggestion. Did you meant to throw an exception while reading the file and let the controller catch and handle it?

adhikasp replied:
Yes something like that, so here you will be "forced" to write the special handling code.
```java 
try {
    List<String> inputData = fileReaderComponent.readFileAsStrings(tapMachineFile);
catch (UnsupportedFile | UnreadableFile e) {
    return false;
}
```

adhikasp commented on src/main/java/com/blibli/future/service/FileReaderComponentImpl.java:
```diff
@@ -0,0 +1,87 @@
+            return readXlsxAsList(file);
+        } else if(filename.endsWith(".xls")) {
+            return readXlsAsList(file);
+        } else {
```
No need for else

riencus replied:
Really? What am I trying to do is to create an error handling when the uploaded file is not a csv or a xls(x) file

adhikasp replied:
```java
} else if(filename.endsWith(".xls")) {
    return readXlsAsList(file);
}
return null
```

adhikasp commented on src/main/java/com/blibli/future/service/FileReaderComponentImpl.java:
```diff
@@ -0,0 +1,87 @@
+            while((line= reader.readLine()) != null) {
+                result.add(line);
+            }
+        } catch (IOException e) {
```
Catching the exception like this is potentially dangerous and could produce empty result if the file corrupt or something thus a 'surprise' result given to user. Maybe `return null` or throw exception that will be handled here https://github.com/blibli-future/athens/pull/1/files#diff-e47c49c4547e3aa56d47d1d685a08224R180

riencus committed:
Modify the files based on adhikasp suggestion

- create new Exception for handling unreadable File
- component -> service
- removing else and change return to throw also lowercase filename
- Modify the controller to adapt the changes in service

riencus committed:
Create Unit test for upload method in controller

Removing null checking because the file is required by the method,
spring will handle it.

riencus committed:
Merge branch 'absensi' into absensi-read-xlsx
